kroxylicious/src/test/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandlerTest.java

```diff
@@ -104,9 +111,73 @@ public static List<Arguments> provideArgsForExpectedFlow() {
         return result;
     }
 
+    /**
+     * If the client sends multiple messages immediately when connecting they may all be read
+     * from the socket despite the inbound channel auto-read being disabled. We need to tolerate
+     * messages being handled while in the CONNECTED/CONNECTING state before the outbound signals
+     * it is active.
+     */
+    @Test
+    void testMultipleMessagesReceivedBeforeOutboundActive() {
```

Comment:
Feels like this test is asserting two or three different behaviours in one go. 

1. The message ordering as seen by the `assertOutboundContainsApiVersionsRequestWithClientSoftwardName` calls. 
2. That the state machine still transitions at the right points to appropriate states
3. That it generates a netFilterContext
4. The channel gets drained at some point (maybe part of 1)

I think those should probably three separate test cases as they are independent reasons to fail tests.

It's not clear to me that the state transitions are tired to the other steps.

Reply:
Pushed up some improvements and splitting, but the nature of the class makes it kind of tricky to design small tests for individual behaviours since you have to drive it through all the transitions to get to your "initial" test state.